### PR TITLE
JIT: Seed stress modes by string hash

### LINKED ISSUE #83733: JIT: Reconsider how stress modes are randomly applied
The hash used to compute whether a stress mode is enabled works simply by xor'ing the integer value of the stress mode in the stress enum:
https://github.com/dotnet/runtime/blob/93dda3be9ce619a0ca58f46ef78f21266aaf42a2/src/coreclr/jit/compiler.cpp#L3581-L3586

That seems like it would have a very bad distribution -- there's only a few dozens of stress modes so this at most affects only the few lowest bits of the hash code, in a very predictable way.

Additionally, reordering members in the enum leads to an entirely different (unrelated) set of stress modes being enabled/disabled. This for example happened in #79283.

I think it would be better if we derived a hash from the string name of each stress mode and then xor'd with that instead (e.g. using `HashStringA`). That would be more similar to how other randomness is mixed in as well.

### PR DESCRIPTION
**Pull request by jakobbotsch**
Instead of using the numeric value of each stress mode inside the enum calculate a hash based on the string name. This has the benefit that it won't change based on new members being added/reordered and reduces correlation of which stress modes get enabled together.

Fix #83733

### CODE REVIEW

**Review by kunalspathak — APPROVED** [review]
LGTM

**Comment by jakobbotsch** [comment]
Looks like this exposes some new stress related issue:

```scala
Assertion failed '(emitThisGCrefRegs & regMask) == 0' in 'System.String:Concat[ushort](System.Collections.Generic.IEnumerable`1[ushort]):System.String' during 'Emit code' 
```

**Comment by jakobbotsch** [comment]
The above issue also hits in libraries-jitstress-random, I opened https://github.com/dotnet/runtime/issues/83963 for it. Will convert this to a draft until we get a fix for that.

File: src/coreclr/jit/compiler.cpp [hunk 1]
```diff
@@ -3491,14 +3498,43 @@ bool Compiler::compStressCompile(compStressArea stressArea, unsigned weight)
     {
         if (verbose)
         {
-            printf("\n\n*** JitStress: %ws ***\n\n", s_compStressModeNames[stressArea]);
+            printf("\n\n*** JitStress: %s ***\n\n", s_compStressModeNames[stressArea]);
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.cpp:3501]: This also fixes this printf for PAL (it was just printing as "JitStress: %ws" on non-Windows builds)

**Comment by jakobbotsch** [comment]
Stress failures should be fixed by #84352, will rerun...

**Comment by jakobbotsch** [comment]
Yay, stress is finally green. Going to remove the stress mode I pointed at in https://github.com/dotnet/runtime/pull/85975#discussion_r1189058870 and then merge.

**Pull request merged**